kernel/device.c

```diff
@@ -177,3 +179,45 @@ void device_busy_clear(struct device *busy_dev)
 	ARG_UNUSED(busy_dev);
 #endif
 }
+
+int device_lock_timeout(struct device *dev, k_timeout_t timeout)
+{
+#ifdef CONFIG_DEVICE_CONCURRENT_ACCESS
+	struct device_context *dc =
+		(struct device_context *)__device_context_start +
+		(dev - __device_start);
+
+	if (k_sem_take(&dc->lock, timeout) != 0) {
+		return -EAGAIN;
+	}
+#endif
+
+	return 0;
+}
+
+void device_call_complete(struct device *dev, int status)
+{
+	struct device_context *dc =
+		(struct device_context *)__device_context_start +
+		(dev - __device_start);
+
+	dc->call_status = status;
+	k_sem_give(&dc->sync);
+}
+
+int device_release(struct device *dev)
+{
+	struct device_context *dc =
+		(struct device_context *)__device_context_start +
+		(dev - __device_start);
+	u32_t status;
+
+	k_sem_take(&dc->sync, K_FOREVER);
```

Comment:
forever here is risky. What if call is never completed.

Reply:
This is improved afterwards with a timeout. However, note that current sync semaphores in drivers afaik always use K_FOREVER. (which is bad indeed)